### PR TITLE
build: Replace CRACO Build/Serve with Vite 

### PR DESCRIPTION
**Pull request by ashtonG**
## Description

This swaps out the utility used to run the dev server and build for the react webapp from craco to vite. Vite has a better track record of keeping up with its underlying dependencies than Craco/Create React App,, and it's build times and dev server startup times are shorter.

## Test Plan

all app functionality should be unchanged.

## Commentary (optional)

Changes to the source code that were made in order to get vite running and building:
* Monaco syntax highlighters and plugins are now directly imported in the codebase rather than managed by plugin
* stray `require` calls are removed
* Notebookjs is now wrapped in a module and imported as a node package
* Custom plugin that fixes a faulty import in the generated swagger api

## Checklist

- [x] Changes have been manually QA'd

## Ticket
[WEB-815]

[WEB-815]: https://determinedai.atlassian.net/browse/WEB-815?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by keita-determined — COMMENTED** [review]
Thats a cool change. Vite is pretty fast, but i found some issues.

- i couldnt launch jupyter with `This 127.0.0.1 page can’t be found` error.
- Every time i run `npm i`, packages-lock is modified. npm version is `9.3.1`

also, i want somebody else from @determined-ai/web to review as well since its a significant change

File: webui/react/src/shared/configs/tsconfig.json [hunk 1]
```diff
@@ -3,7 +3,7 @@
     "target": "es2020",
     "lib": ["dom", "dom.iterable", "esnext"],
     "allowJs": false,
-    "skipLibCheck": false,
+    "skipLibCheck": true,
```
[the remark this answers is not in the document]
Comment (ashtonG) [reply webui/react/src/shared/configs/tsconfig.json:6]: This option was actually overridden in create react app to be false, so this is more restoring functionality that already existed. By enabling the library check, typescript throws an error based on the types in react.

File: webui/react/package.json [hunk 2]
```diff
@@ -8,8 +8,8 @@
   },
   "scripts": {
     "postshrinkwrap": "npx npm-force-resolutions && npm install --ignore-scripts",
-    "start": "craco start",
-    "build": "craco build",
+    "start": "vite serve",
```
[the remark this answers is not in the document]
Comment (ashtonG) [reply webui/react/package.json:11]: As mentioned in the engineering doc and meeting, vite intentionally doesn't support the stuff that would make the type checking in create react app work. I don't think this is a terrible loss because most modern editors are able to leverage the preexisting typescript language server to enable type checking. We do manually check types during the build just to be safe, which should cause the react build to fail during github ci.

just pushed an update to ensure the server starts on port 3000.

[the remark this answers is not in the document]
Comment (keita-determined) [reply webui/react/package.json:11]: i think its good to have type check. 
https://vitejs.dev/guide/features.html#transpile-only

>During development, if you need more than IDE hints, we recommend running tsc --noEmit --watch in a separate process, or use [vite-plugin-checker](https://github.com/fi3ework/vite-plugin-checker) if you prefer having type errors directly reported in the browser.

File: webui/react/src/shared/configs/tsconfig.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (keita-determined) [reply webui/react/src/shared/configs/tsconfig.json:6]: ah does the error come from table component? if so, that makes sense. we need to fix it at some point

[the remark this answers is not in the document]
Comment (ClaireNeveu) [reply webui/react/src/shared/configs/tsconfig.json:6]: Can we just fix that type error?

[the remark this answers is not in the document]
Comment (ashtonG) [reply webui/react/src/shared/configs/tsconfig.json:6]: @keita-determined no, the errors are in the type definitions. You can see this for yourself by changing `skipLibCheck` to `false` locally and running `npx tsc`.

[the remark this answers is not in the document]
Comment (ClaireNeveu) [reply webui/react/src/shared/configs/tsconfig.json:6]: Is that because the version of the type definitions and react library are mismatched?

[the remark this answers is not in the document]
Comment (ashtonG) [reply webui/react/src/shared/configs/tsconfig.json:6]: @ClaireNeveu No -- we've configured typescript not to check the type of plain javascript modules. These errors are in the type definitions themselves.

**Comment by ashtonG** [comment]
@keita-determined I'm unable to reproduce either issue locally or on the netlify build. Can you please verify your setup?

**Comment by keita-determined** [comment]
> @keita-determined I'm unable to reproduce either issue locally or on the netlify build. Can you please verify your setup?

- i couldnt launch jupyter with This 127.0.0.1 page can’t be found error.
  - This one is fixed with [this change](https://github.com/determined-ai/determined/pull/5926/commits/15a1bd3009426b4c72697e91c596ef1e8b59b3d2)
- Every time i run npm i, packages-lock is modified. npm version is 9.3.1
  - I still get this one. I just ran the `npm i`. node version is `v16.16.0` , npm version is 9.3.1
- Also, `make live` opens `http://127.0.0.1:3000`, but preferably `localhost:3000` is ideal.

File: webui/react/package.json [hunk 3]
```diff
@@ -8,8 +8,8 @@
...
+    "build": "tsc; NODE_OPTIONS=--max-old-space-size=6144 vite build --mode=${DET_NODE_ENV:-production}",
```
[the remark this answers is not in the document]
Comment (ashtonG) [reply webui/react/package.json:12]: Yes -- the bundler that vite uses is less well behaved than webpack when it comes to memory usage. We're using 6gb in particular because our ci container allocates 8gb -- while the program should gracefully allocate extra memory to disk if `max-old-space` exceeds the physical memory of the container, that's slow.

File: webui/react/src/shared/utils/datetime.ts [hunk 4]
```diff
@@ -1,12 +1,15 @@
...
 
+// TODO: check types for this.
+// @ts-ignore
+import humanizeDuration from 'humanize-duration';
+
 import { StartEndTimes } from 'types';
 
 dayjs.extend(utc);
 
 /* eslint-disable @typescript-eslint/no-var-requires */
```
Comment (hkang1) [inline webui/react/src/shared/utils/datetime.ts]: question: is this line still needed with the removal of the "require"? awesome that we can conform to `import`!

File: webui/react/src/shared/utils/datetime.ts [hunk 5]
```diff
@@ -1,12 +1,15 @@
 import dayjs from 'dayjs';
 import utc from 'dayjs/plugin/utc';
 
+// TODO: check types for this.
+// @ts-ignore
```
Comment (hkang1) [inline webui/react/src/shared/utils/datetime.ts]: chore: just adding as a reminder if you wanted to address this before PR wrapup

File: webui/react/src/shared/utils/dom.ts [hunk 6]
```diff
@@ -1,5 +1,5 @@
 /* eslint-disable @typescript-eslint/no-var-requires */
```
Comment (hkang1) [inline webui/react/src/shared/utils/dom.ts]: question: ditto for require

File: webui/react/.eslintignore [hunk 7]
```diff
@@ -1,6 +1,7 @@
 src/serviceWorker.ts
 src/services/api-ts-sdk/
 src/vendor/notebook.js
```
Comment (hkang1) [inline webui/react/.eslintignore:3]: chore: since it moved, can we drop this line?

File: webui/react/Makefile [hunk 8]
```diff
@@ -22,7 +23,7 @@ get-deps: check-requirements node_modules/done.stamp
 .PHONY: build
 build: check-requirements build/done.stamp
 
-build/done.stamp: $(source_files) node_modules/done.stamp tsconfig.json craco.config.js jest.config.js
+build/done.stamp: $(source_files) node_modules/done.stamp tsconfig.json craco.config.js jest.config.js vite.config.ts
```
Comment (hkang1) [inline webui/react/Makefile:26]: question: probably part of the plan but just wanting to confirm that we are removing craco from this PR. `craco.config.js`, craco packages, etc

File: webui/react/src/pages/ExperimentDetails/CodeViewer/IpynbRenderer.tsx [hunk 9]
```diff
@@ -1,7 +1,6 @@
+import NotebookJS from 'notebook';
```
Comment (hkang1) [inline webui/react/src/pages/ExperimentDetails/CodeViewer/IpynbRenderer.tsx:1]: question: not part of this PR, but question for @EmilyBonar (I think): there must be a reason but curious as to why `notebook.js` can't just be in the `package.json` and call it a day

**Review by hkang1 — COMMENTED** [review]
Overall looked great! Sprinkled comments throughout based on the changes, but I'll run this branch once I have everything setup properly on my machine to actually run these changes.

**Comment by hkang1** [comment]
Tested:
- `make live` - works as expected, but noticed that I generally had to refresh a couple of times before it took me to the login page? Jupyterlab does not run as it can't resolve via localhost:<live port>. Not sure if this ever ran properly via webdevserver though.
- `make build` - works as expected locally (via devclusteer), jupyterlab runs perfectly

File: webui/react/src/pages/ExperimentDetails/CodeViewer/IpynbRenderer.tsx (hunk 9 above)
Comment (ashtonG) [reply webui/react/src/pages/ExperimentDetails/CodeViewer/IpynbRenderer.tsx:1]: We vendored `notebook.js` because the bundler includes `jsdom` in our bundle, so the vendored version now assumes it's running in a browser.

File: webui/react/src/shared/configs/tsconfig.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (ashtonG) [reply webui/react/src/shared/configs/tsconfig.json:6]: @ClaireNeveu after poking at this for a few hours, I'm going to open a ticket for this task and push to handle this as a fast follow because getting the library types to work all the way through will cause a large expansion in scope. I was able to get most of the errors to clear by updating libraries, but there's a remaining issue in `@types/node` stemming from a difference in URL objects between browsers and nodejs that the swagger api library is bumping into. We might need to either do a complex find/replace as part of the binding generation or implement a custom codegen class, which is doable, but I'd want that change to be looked at separate from the other build issues.

File: webui/react/src/shared/utils/datetime.ts (hunk 5 above)
Comment (ashtonG) [reply webui/react/src/shared/utils/datetime.ts]: This was changed in a future commit

File: webui/react/.eslintignore (hunk 7 above)
Comment (ashtonG) [reply webui/react/.eslintignore]: done!

File: webui/react/Makefile (hunk 8 above)
Comment (ashtonG) [reply webui/react/Makefile:26]: We're keeping CRACO for now because it's handling our tests. While we could migrate to Vitest or ava or stay with jest, changing the testing framework would require modifying all of our preexisting tests (even if we ejected from craco and stayed with jest, we'd likely want to ensure that it used the transpilation library we use in our build, which would require changes to our tests due to mocking behavior changes stemming from the difference between es modules and commonjs modules).

Comment (hkang1) [reply webui/react/Makefile:26]: thanks makes total sense

**Comment by ashtonG** [comment]
@hkang1 when you say jupyterlab doesn't work, do you mean that it's stuck on queued or that the interactive page serves a 404? I just pushed a change that seems to prevent the popup url from being malformed. Can you try it out?

@keita-determined i think this might be the same issue you pointed out earlier.

**Review by hkang1 — APPROVED** [review]
All the comments and questions I had have been addressed. Conditional approval pending other folks' inquiries and comments.

**Review by keita-determined — COMMENTED** [review]
overall LGTM too

few notes:

- `make live` opens `http://127.0.0.1:3000/det` instead of `localhost`. According to [this](https://github.com/vitejs/vite/issues/9195#issuecomment-1188596943), its intended, so i think its good with `http://127.0.0.1:3000/det`. if there is an issue with this, we can change it in the future.

---

- I still get the `packages-lock.json` change after `npm i` with node `v16.16.0` and npm `9.3.1`, but i can easily revert this change, so its ok.
<details>
<summary>Diff I got</summary>

```diff
diff --git a/webui/react/package-lock.json b/webui/react/package-lock.json
index 47ae0321f..e95e161f3 100644
--- a/webui/react/package-lock.json
+++ b/webui/react/package-lock.json
@@ -17999,8 +17999,13 @@
       }
     },
     "node_modules/notebook": {
-      "resolved": "src/vendor/notebook",
-      "link": true
+      "version": "0.7.0",
+      "resolved": "file:src/vendor/notebook",
+      "dependencies": {
+        "ansi_up": "^5.1.0",
+        "dompurify": "^2.4.0",
+        "marked": "^4.1.1"
+      }
     },
     "node_modules/npm-force-resolutions": {
       "version": "0.0.10",
@@ -28572,14 +28577,6 @@
         "type": "github",
         "url": "https://github.com/sponsors/wooorm"
       }
-    },
-    "src/vendor/notebook": {
-      "version": "0.7.0",
-      "dependencies": {
-        "ansi_up": "^5.1.0",
-        "dompurify": "^2.4.0",
-        "marked": "^4.1.1"
-      }
     }
   },
   "dependencies": {
@@ -40950,7 +40947,7 @@
       }
     },
     "notebook": {
-      "version": "file:src/vendor/notebook",
+      "version": "0.7.0",
       "requires": {
         "ansi_up": "^5.1.0",
         "dompurify": "^2.4.0",
```
</details>

---

nit: `npm run analyze` is not working on my local. i wonder how many ppl use this command
this issue can be fixed in diff PR
